src/rpp/rpp/operators/combine_latest.hpp

```diff
@@ -0,0 +1,166 @@
+            }
+        }
+
+        void offer_completed(const auto& subscriber)
+        {
+            if (!subscriber.is_subscribed()) return;
+
+            std::scoped_lock<std::mutex> lock{m_mutex};
+
+            if (++m_completed_count == m_total_completed_count)
+            {
+                subscriber.on_completed();
+            }
+        }
+
+    private:
+        TCombiner m_combiner;
+
+        std::mutex m_mutex;
+        std::tuple<std::optional<SourceValueType>, std::optional<OtherValueTypes>...> m_values{};
+        size_t m_completed_count{0};
+
+        const std::size_t m_total_completed_count = 1 + sizeof...(OtherValueTypes);
+
+        bool has_values_for_observer()
+        {
+            return std::apply([](const auto&... cached_values)
+            {
+                // Variadic expansion used
+                return (cached_values.has_value() && ...);
+            }, m_values);
+        }
+
+        auto combine_cached_values()
+        {
+            return std::apply([&](const auto&... cached_values)
+            {
+                // Variadic expansion used
+                return m_combiner(cached_values.value()...);
+            }, m_values);
+        }
+    };
+
+    /**
+     * \brief Templated helper function for subscribing to variadic 'other' observables.
+     *
+     * \param observables is a variadic tuple. The first observable is the current observable and rest are the other observables.
+     * \param subscriber is the downstream subscriber.
+     * \param coordinator manages the cache of emission from the observables and coordinate dispatching.
+     */
+    template<size_t...I>
+    static void subscribe_other_observables(std::index_sequence<I...>, // Used in compile time for variadic expansion
+                                            const auto& observables,
+                                            const auto& subscriber,
+                                            const auto& coordinator)
+    {
+        // +1 because the first element in tuple is the current observable, and you want to subscribe to the 'other' observables.
+        // (Use variadic expansion to iterate the observables)
+        (subscribe_observable<I + 1>(std::get<I>(observables), subscriber, coordinator), ...);
+    }
+
+    template<size_t I, constraint::observable TObservable>
+    static void subscribe_observable(const TObservable& observable, const auto& subscriber, const auto& coordinator)
+    {
+        using ValueType = utils::extract_observable_type_t<TObservable>;
+        observable.subscribe(create_inner_subscriber<ValueType, I>(subscriber, coordinator));
+    }
+
+    template<typename ValueType, size_t I>
+    static auto create_inner_subscriber(auto&& subscriber, const auto& coordinator)
+    {
+        return create_subscriber_with_state<ValueType>(
+                    subscriber.get_subscription().make_child(),
+                    std::forward<decltype(subscriber)>(subscriber),
```

Comment:
Error-prone line! extract getting subscription as separate line and then move it there. What if subscriber was moved via `forward` before getting subscription???

Reply:
Oh you're right. I was thinking if this will work last night too.

Reply:
No, the order of evaluation of function arguments is unspecified. As a result, std::forward can happens BEFORE getting of the subscription

Reply:
Simple example on [godbolt](https://godbolt.org/z/Mbv9WEodK)

```cpp
#include <iostream>

struct A
{
    A(int i) { std::cout << "Constructed A " << i << std::endl; }
};

void f(A a1, A a2, A a3){}

int main()
{
    f(A{1}, A{2}, A{3});
    return 0;
}
```
MSVC/GCC
```
Constructed A 3
Constructed A 2
Constructed A 1
```
Clang
```
Constructed A 1
Constructed A 2
Constructed A 3
```

Reply:
Wow, this is really interesting!